File: scribe.py
import sys, os, glob
from functools import partial
import xxhash
# import rapidjson as json
from datetime import datetime
import sqlite3
# import concurrent.futures


def init_db():
    cur.execute("""CREATE TABLE IF NOT EXISTS hashes
                        (name TEXT, filesize INTEGER, hash TEXT, date TEXT)""")

    cur.execute("""CREATE TABLE IF NOT EXISTS files
                        (name TEXT, filesize INTEGER, date TEXT)""")

    cur.execute("""CREATE TABLE IF NOT EXISTS dates
                        (date TEXT)""")

    cur.execute("""CREATE INDEX IF NOT EXISTS hashes_name_idx
                        ON hashes(name)""")

    cur.execute("""CREATE INDEX IF NOT EXISTS files_name_idx
                        ON files(name)""")

    cur.execute("""CREATE INDEX IF NOT EXISTS files_date_idx
                        ON files(date)""")

    cur.execute("""CREATE INDEX IF NOT EXISTS files_name_date_idx
                        ON files(name, date)""")


SEED=153
COMMIT_RATE = 300

path = sys.argv[1]
num_threads = 1

db = sqlite3.connect(f"{path}/.scribe.db")
cur = db.cursor()
init_db()
insert_count = 0

# pool = concurrent.futures.ThreadPoolExecutor(max_workers=num_threads)

def get_hash(file):
    CHUNK_SIZE = 2 ** 17  # Or whatever you have memory to handle
    with open(file, 'rb') as input_file:
        x = xxhash.xxh3_64(seed=SEED)
        for chunk in iter(partial(input_file.read, CHUNK_SIZE), b''):
            x.update(chunk)
        return x.hexdigest()

def write_hash(f, datetime, force_hash=False):
    global insert_count
    try:
        filesize = os.stat(f).st_size

        cur.execute('SELECT name, filesize FROM files WHERE name = ? ORDER BY date DESC LIMIT 1', (f, ))
        res = cur.fetchone()


        h = ""
        if not res:
            h = get_hash(f)
        elif res[1] != filesize:
            h = get_hash(f)

        d = None
        if h == "":
            d = (f,filesize, datetime,)
            cur.execute("INSERT INTO files(name, filesize, date) VALUES(?,?,?)", (f,filesize, datetime))
        else:
            d = (f,filesize, datetime,h,)
            cur.execute("INSERT INTO files(name, filesize, date) VALUES(?,?,?)", (f,filesize, datetime))
            cur.execute("INSERT INTO hashes(name, filesize, date, hash) VALUES(?,?,?,?)", (f,filesize, datetime, h))

        # print(f"{f=}")
        # print(f"{h=}")

        insert_count += 1
        if not insert_count % COMMIT_RATE:
            db.commit()
            insert_count = 0


    except FileNotFoundError as e:
        print(e)



def get_datetime():
    today = datetime.now()
    date_time_str = today.strftime("%Y-%m-%d %H:%M:%S")
    return date_time_str

def write_hashes():
    datetime = get_datetime()
    count = 0
    for subdir, dirs, files in os.walk(path):
        for file in files:
            f = subdir + os.sep + file
            if os.path.isdir(f): continue
            if f.endswith(".scribe"): continue

            # print(f)
            write_hash(f, datetime)
            count += 1


write_hashes()
db.commit()
db.close()

File: scribe.nim
import std/[strutils, streams, os, osproc,
            strformat, times, paths, math]
import print
import xxhash
import db_connector/db_sqlite


let seed = 153.uint64

let scribeDir = (if paramCount() > 0 : paramStr(1) else: "./")

let dbPath = fmt"{scribeDir}/.scribe2.db"
let db = open(dbPath, "", "", "")

proc initDb(): void =
  db.exec(sql"""CREATE TABLE IF NOT EXISTS hashes
                      (name TEXT, filesize INTEGER, hash TEXT, date TEXT)""")

  db.exec(sql"""CREATE TABLE IF NOT EXISTS files
                      (name TEXT, filesize INTEGER, date TEXT)""")

  db.exec(sql"""CREATE TABLE IF NOT EXISTS dates
                      (date TEXT)""")

  db.exec(sql"""CREATE INDEX IF NOT EXISTS hashes_name_idx
                      ON hashes(name)""")

  db.exec(sql"""CREATE INDEX IF NOT EXISTS files_name_idx
                      ON files(name)""")

  db.exec(sql"""CREATE INDEX IF NOT EXISTS files_date_idx
                      ON files(date)""")

  db.exec(sql"""CREATE INDEX IF NOT EXISTS files_name_date_idx
                      ON files(name, date)""")



proc shouldIgnore(path: string): bool =
  let ignorePaths = @[".scribe", ".vim", ".git", "."]
  for pathPart in path.split('/'):

    for p in ignorePaths:
      if p.len == 1: continue

      if pathPart.endsWith(p):
        return true

      if pathPart.startsWith(p):
        return true

  return false


const chunkSize = 2 ^ 17
var buffer = newSeqUninitialized[uint8](chunkSize)
# var buffer = newString(chunkSize)

proc streamingHash(path: string): string =
  var fs = newFileStream(path, fmRead)
  defer: fs.close()

  var state = newXxh3_64(seed)
  var size: int

  while not fs.atEnd:
    # fs.readStr(chunkSize, buffer)
    let size = fs.readData(addr(buffer[0]), chunkSize - 1)
    buffer[size] = 0

    state.update(cast[string](buffer))

  return state.digest().toHex.toLowerAscii



var insertCount = 0
let COMMIT_RATE = 150

var dateStr = now().format("yyyy-MM-dd' 'HH:mm:ss")
var count = 0
var txStarted = false
proc hashFiles() =
  for p in walkDirRec(scribeDir & "/"):

    if p.shouldIgnore:
      continue

    let path = p.normalizedPath
    insertCount += 1
    # if insertCount mod 1000 == 0:
    #   print path

    let filesize = path.getFileSize

    let fetchedSize = db.getValue(sql"""SELECT filesize FROM files
                                      WHERE name = ? ORDER BY date
                                      DESC LIMIT 1""", path)


    var hash = ""
    if fetchedSize.len == 0 or fetchedSize.parseInt != filesize:
      hash = streamingHash(path)

    if not txStarted:
      db.exec(sql"BEGIN")
      txStarted = true

    if hash == "":
      db.exec(sql"INSERT INTO files(name, filesize, date) VALUES(?,?,?)",
                 path,filesize, dateStr)
    else:
      db.exec(sql"INSERT INTO files(name, filesize, date) VALUES(?,?,?)",
                 path,filesize, dateStr)

      db.exec(sql"INSERT INTO hashes(name, filesize, date, hash) VALUES(?, ?,?,?)",
                 path,filesize, dateStr, hash)
    count += 1

    if count mod 300 == 0:
      db.exec(sql"COMMIT")
      print "commit"
      txStarted = false
      count = 0

    # print path
    # print hash
  if txStarted:
    db.exec(sql"COMMIT")

proc checkFiles() =
  let lastTime = db.getValue(sql"""SELECT date FROM files
                                    ORDER BY date
                                    DESC LIMIT 1""")

  for row in db.rows(sql"SELECT files.name, hashes.hash FROM files JOIN hashes ON files.name = hashes.name WHERE files.date = ?", lastTime):
    if not row[0].fileExists:
      print "not found"
      print row[0], row[1]
    else:
      let newHash = streamingHash(row[0])
      print newHash, row[1]
      if newHash == row[1]:
        print "match"
      else:
        print row[0]
        print "no match ERROROROROROR"
        discard readline stdin






initDb()
# hashFiles()
# checkFiles()
db.close()
